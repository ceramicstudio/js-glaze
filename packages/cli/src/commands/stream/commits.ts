import { SyncOptions } from '@ceramicnetwork/common'
import { 
  Command, 
  STREAM_ID_ARG,
  SYNC_OPTION_FLAG,
  type CommandFlags
} from '../../command.js'

type Flags = CommandFlags & {
  syncOption?: SyncOptions
}

export default class Commits extends Command<Flags, { streamId: string }> {
  static description = 'list commits contained within a stream'

  static args = [
    STREAM_ID_ARG,
  ]

  static flags = {
    ...Command.flags,
    syncOption: SYNC_OPTION_FLAG,
  }

  async run(): Promise<void> {
    this.spinner.start(`Loading stream ${this.args.streamId}...`)
    try {
      const stream = await this.ceramic.loadStream(
        this.args.streamId,
        { sync: this.flags.syncOption }
      )
      const commits = stream.allCommitIds.map((v) => v.toString())
      this.spinner.succeed(`Stream commits loaded.`)
      this.logJSON(commits)
    } catch (e) {
      this.spinner.fail((e as Error).message)
    }
  }
}
