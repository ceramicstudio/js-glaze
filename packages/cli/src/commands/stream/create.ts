import { flags } from '@oclif/command'
import { TileDocument } from '@ceramicnetwork/stream-tile'

import { Command } from '../../command'
import type { CommandFlags } from '../../command'
import { parseContent, parseControllers } from '../../utils'

type Flags = CommandFlags & {
  determinitic?: boolean
  'only-genesis'?: boolean

  did?: string
  controllers?: string
}

export default class Create extends Command<
  Flags,
  {
    schema: string
    content: string
  }
> {
  static description = 'create a new stream'

  static args = [
    {
      name: 'schema',
      description: 'the StreamId of the schema to use',
      required: true,
    },
    {
      name: 'content',
      description: 'the Stream body',
      required: true,
    },
  ]

  static flags = {
    ...Command.flags,
    'only-genesis': flags.boolean({
      char: 'g',
      description: 'only generate genesis block',
      default: false,
    }),
    deterministic: flags.boolean({
      char: 'd',
      description: 'generate deterministic stream',
      default: false,
    }),

    did: flags.string({
      exclusive: ['key'],
      description: 'Creator DID',
    }),
    controllers: flags.string({
      char: 'c',
      description: 'Comma separated list of controllers',
    }),
  }

  async run(): Promise<void> {
    this.spinner.start('Creating stream...')
    let did: string | undefined
    if (this.flags.key != null) {
      did = this.authenticatedDID.id
    } else if (this.flags.did !== null) {
      did = this.flags.did
    } else {
      throw new Error('Missing DID')
    }
    try {
      let parsedControllers: Array<string>
      if (this.flags.controllers !== undefined) {
        parsedControllers = parseControllers(this.flags.controllers)
      } else if (did !== undefined) {
        parsedControllers = parseControllers(did)
      } else {
        throw new Error('No DID to assign as a controller')
      }
      const parsedContent = parseContent(this.args.content)

      const localDeterministic =
        this.flags.deterministic !== undefined ? this.flags.determinitic : false

      const metadata = {
        controllers: parsedControllers,
        schema: this.args.schema,
        deterministic: localDeterministic,
      }

      const tile = await TileDocument.create(this.ceramic, parsedContent, metadata, {
        anchor: !this.flags['only-genesis'],
        publish: !this.flags['only-genesis'],
      })

      this.spinner.succeed(`Created Stream ${tile.commitId.toString()}.`)
      this.logJSON({
        streamID: tile.id.toString(),
        commitId: tile.commitId.toString(),
        body: tile.content,
      })
    } catch (e) {
      console.error(e)
      this.spinner.fail((e as Error).message)
      return
    }
  }
}
