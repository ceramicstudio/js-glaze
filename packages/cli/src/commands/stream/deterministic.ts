import { flags } from '@oclif/command'
import { TileDocument } from '@ceramicnetwork/stream-tile'
import type { TileMetadataArgs } from '@ceramicnetwork/stream-tile'

import { Command } from '../../command'
import type { CommandFlags } from '../../command'

type Flags = CommandFlags & {
  metadata?: TileMetadataArgs
}

export default class Deterministic extends Command<
  Flags,
  {
    content: string
  }
> {
  static description = 'Create a new deterministic Stream'

  static args = [
    {
      name: 'content',
      description: 'The contents of the stream',
      required: true,
    },
  ]

  static flags = {
    ...Command.flags,
    metadata: flags.string({
      char: 'm',
      description: 'Stream Metadata',
      parse: JSON.parse,
    }),
  }

  async run(): Promise<void> {
    this.spinner.start('Creating stream...')
    let did: string | undefined
    if (this.flags.key != null) {
      did = this.authenticatedDID.id
    } else {
      throw new Error('No DID cached, please provide your key.')
    }
    try {
      const metadata = {
        ...this.flags.metadata,
        deterministic: true,
      } || {
        controllers: [did],
        deterministic: true,
      }

      metadata?.controllers ? undefined : (metadata.controllers = [did])

      const tile = await TileDocument.create(this.ceramic, this.args.content, metadata)

      this.spinner.succeed(`Created Stream ${tile.commitId.toString()}.`)
      this.logJSON({
        streamID: tile.id.toString(),
        commitId: tile.commitId.toString(),
        body: tile.content,
      })
    } catch (e) {
      console.error(e)
      this.spinner.fail((e as Error).message)
      return
    }
  }
}
