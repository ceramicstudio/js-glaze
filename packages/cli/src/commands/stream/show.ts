import { TileDocument } from '@ceramicnetwork/stream-tile'
import StreamID from '@ceramicnetwork/streamid'

import { Command } from '../../command'
import type { CommandFlags } from '../../command'

export default class Show extends Command<
  CommandFlags,
  {
    streamId: string
  }
> {
  static description = 'Show content of a stream'
  static args = [{ name: 'streamId', description: 'StreamID to be queried' }]

  async run(): Promise<void> {
    this.spinner.start(`Querying stream ${this.args.streamId}`)
    try {
      const stream = await TileDocument.load(this.ceramic, StreamID.fromString(this.args.streamId))
      this.spinner.succeed('Stream details retrieved.')
      this.logJSON(stream.content)
    } catch (e) {
      this.spinner.fail((e as Error).message)
      throw e
    }
  }
}
