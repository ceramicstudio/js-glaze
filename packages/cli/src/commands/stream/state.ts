import { Command } from '../../command'
import type { CommandFlags } from '../../command'

export default class State extends Command<CommandFlags, { streamId: string }> {
  static description = 'get the state of a Stream'

  static args = [
    {
      name: 'streamId',
      required: true,
      description: 'ID of the Stream',
    },
  ]

  async run(): Promise<void> {
    this.spinner.start(`Querying stream ${this.args.streamId}`)

    try {
      const stream = await this.ceramic.loadStream(this.args.streamId)
      this.spinner.succeed(`Successfully queried stream ${this.args.streamId}`)
      this.logJSON(stream)
    } catch (e) {
      this.spinner.fail((e as Error).message)
    }
  }
}
