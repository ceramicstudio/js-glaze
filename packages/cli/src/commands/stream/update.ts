import { flags } from '@oclif/command'
import { TileDocument } from '@ceramicnetwork/stream-tile'

import { Command } from '../../command'
import type { CommandFlags } from '../../command'
import { parseControllers, parseContent } from '../../utils'

type Flags = CommandFlags & {
  controllers?: string
  did?: string
}

export default class Update extends Command<
  Flags,
  {
    streamId: string
    content: string
  }
> {
  static description = 'Update a stream'

  static args = [
    { name: 'streamId', description: 'Document StreamID', required: true },
    { name: 'content', description: 'Document Content', required: true },
  ]
  static flags = {
    ...Command.flags,

    did: flags.string({
      exclusive: ['key'],
      description: 'Creator DID',
    }),
    controllers: flags.string({
      char: 'c',
      description: 'Comma separated list of controllers',
    }),
  }

  async run(): Promise<void> {
    this.spinner.start('Updating stream...')

    let did: string | undefined
    if (this.flags.key != null) {
      did = this.authenticatedDID.id
    } else if (this.flags.did !== null) {
      did = this.flags.did
    } else {
      throw new Error('Missing DID')
    }
    try {
      const parsedContent = parseContent(this.args.content)
      let parsedControllers: Array<string>
      if (this.flags.controllers !== undefined) {
        parsedControllers = parseControllers(this.flags.controllers)
      } else if (did !== undefined) {
        parsedControllers = parseControllers(did)
      } else {
        throw new Error('No DID to assign as a controller')
      }

      const doc = await TileDocument.load(this.ceramic, this.args.streamId)

      const metadata = {
        controllers: parsedControllers,
      }

      await doc.update(parsedContent, metadata)
      this.spinner.succeed('Updated stream')
      this.logJSON({ commitId: doc.commitId.toString(), content: doc.content })
    } catch (e) {
      this.spinner.fail((e as Error).message)
    }
  }
}
