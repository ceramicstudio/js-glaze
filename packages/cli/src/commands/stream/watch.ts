import chalk from 'chalk'
import { TileDocument } from '@ceramicnetwork/stream-tile'

import { Command } from '../../command'
import type { CommandFlags } from '../../command'

export default class Watch extends Command<
  CommandFlags,
  {
    streamId: string
  }
> {
  static description = 'Monitor stream for any updates.'
  static args = [{ name: 'streamId', required: true, description: 'Stream ID' }]

  async run(): Promise<void> {
    this.spinner.start('Finding Stream...')
    try {
      const doc = await TileDocument.load(this.ceramic, this.args.streamId)
      doc.subscribe(() => {
        console.log(chalk.green('--- Document updated ---'))
        this.logJSON(doc.content)
      })
    } catch (e) {
      this.spinner.fail((e as Error).message)
    }
  }
}
