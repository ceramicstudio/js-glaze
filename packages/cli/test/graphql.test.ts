import { execa } from 'execa'
import { Readable } from 'node:stream'

describe('graphql', () => {
  describe('graphql:schema', () => {
    test('printing graphql schema succeeds', async () => {
      const schema = await execa('glaze', [
        'graphql:schema',
        'test/mocks/runtime.composite.picture.post.json',
      ])
      expect(schema.stdout.toString()).toMatchSnapshot()
    }, 60000)

    test('printing graphql schema succeeds with --readonly flag', async () => {
      const schema = await execa('glaze', [
        'graphql:schema',
        'test/mocks/runtime.composite.picture.post.json',
        '--readonly',
      ])
      expect(schema.stdout.toString()).toMatchSnapshot()
    }, 60000)
  })

  describe('graphql:server', () => {
    test('graphql server starts', async () => {
      const serverProcess = execa('glaze', [
        'graphql:server',
        'test/mocks/runtime.composite.picture.post.json',
        '--port=62433',
      ])
      let numChecks = 0
      serverProcess.stdout?.on('data', (data: Readable) => {
        if (numChecks === 0) {
          expect(
            data
              .toString()
              .includes('GraphQL server is listening on http://localhost:62433/graphql')
          ).toBe(true)
          numChecks++
        } else if (numChecks === 1) {
          expect(data.toString().includes('Server stopped')).toBe(true)
        }
      })
      await Promise.race([
        new Promise((resolve) =>
          setTimeout(() => {
            serverProcess.kill()
            resolve(true)
          }, 10000)
        ),
        serverProcess,
      ])
    }, 60000)

    test('graphql server starts with --readonly flag', async () => {
      const serverProcess = execa('glaze', [
        'graphql:server',
        'test/mocks/runtime.composite.picture.post.json',
        '--port=62610',
        '--readonly',
      ])
      let numChecks = 0
      serverProcess.stdout?.on('data', (data: Readable) => {
        if (numChecks === 0) {
          expect(
            data
              .toString()
              .includes('GraphQL server is listening on http://localhost:62610/graphql')
          ).toBe(true)
          numChecks++
        } else if (numChecks === 1) {
          expect(data.toString().includes('Server stopped')).toBe(true)
        }
      })
      await Promise.race([
        new Promise((resolve) =>
          setTimeout(() => {
            serverProcess.kill()
            resolve(true)
          }, 10000)
        ),
        serverProcess,
      ])
    }, 60000)
  })
})
