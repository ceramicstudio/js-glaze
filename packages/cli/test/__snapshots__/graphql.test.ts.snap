// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql graphql:schema printing graphql schema succeeds 1`] = `
"type Query {
  \\"\\"\\"Fetches an object given its ID\\"\\"\\"
  node(
    \\"\\"\\"The ID of an object\\"\\"\\"
    id: ID!
  ): Node
  account(id: DID!): CeramicAccount!
  viewer: CeramicAccount
}

\\"\\"\\"An object with an ID\\"\\"\\"
interface Node {
  \\"\\"\\"The id of the object.\\"\\"\\"
  id: ID!
}

type CeramicAccount {
  id: DID!
  isViewer: Boolean!
  data: CeramicAccountData!
}

\\"\\"\\"
A field whose value conforms to the standard DID format as specified in did-core: https://www.w3.org/TR/did-core/.
\\"\\"\\"
scalar DID @specifiedBy(url: \\"https://www.w3.org/TR/did-core/\\")

type CeramicAccountData {
  pictureCollection(
    \\"\\"\\"Returns the items in the list that come after the specified cursor.\\"\\"\\"
    after: String

    \\"\\"\\"Returns the first n items from the list.\\"\\"\\"
    first: Int

    \\"\\"\\"Returns the items in the list that come before the specified cursor.\\"\\"\\"
    before: String

    \\"\\"\\"Returns the last n items from the list.\\"\\"\\"
    last: Int
  ): PictureConnection
  postCollection(
    \\"\\"\\"Returns the items in the list that come after the specified cursor.\\"\\"\\"
    after: String

    \\"\\"\\"Returns the first n items from the list.\\"\\"\\"
    first: Int

    \\"\\"\\"Returns the items in the list that come before the specified cursor.\\"\\"\\"
    before: String

    \\"\\"\\"Returns the last n items from the list.\\"\\"\\"
    last: Int
  ): PostConnection
}

\\"\\"\\"A connection to a list of items.\\"\\"\\"
type PictureConnection {
  \\"\\"\\"Information to aid in pagination.\\"\\"\\"
  pageInfo: PageInfo!

  \\"\\"\\"A list of edges.\\"\\"\\"
  edges: [PictureEdge]
}

\\"\\"\\"Information about pagination in a connection.\\"\\"\\"
type PageInfo {
  \\"\\"\\"When paginating forwards, are there more items?\\"\\"\\"
  hasNextPage: Boolean!

  \\"\\"\\"When paginating backwards, are there more items?\\"\\"\\"
  hasPreviousPage: Boolean!

  \\"\\"\\"When paginating backwards, the cursor to continue.\\"\\"\\"
  startCursor: String

  \\"\\"\\"When paginating forwards, the cursor to continue.\\"\\"\\"
  endCursor: String
}

\\"\\"\\"An edge in a connection.\\"\\"\\"
type PictureEdge {
  \\"\\"\\"The item at the end of the edge\\"\\"\\"
  node: Picture

  \\"\\"\\"A cursor for use in pagination\\"\\"\\"
  cursor: String!
}

type Picture implements Node {
  id: ID!
  url: String!
}

\\"\\"\\"A connection to a list of items.\\"\\"\\"
type PostConnection {
  \\"\\"\\"Information to aid in pagination.\\"\\"\\"
  pageInfo: PageInfo!

  \\"\\"\\"A list of edges.\\"\\"\\"
  edges: [PostEdge]
}

\\"\\"\\"An edge in a connection.\\"\\"\\"
type PostEdge {
  \\"\\"\\"The item at the end of the edge\\"\\"\\"
  node: Post

  \\"\\"\\"A cursor for use in pagination\\"\\"\\"
  cursor: String!
}

type Post implements Node {
  id: ID!
  text: String!
}

type Mutation {
  createPicture(input: CreatePictureInput!): CreatePicturePayload
  updatePicture(input: UpdatePictureInput!): UpdatePicturePayload
  createPost(input: CreatePostInput!): CreatePostPayload
  updatePost(input: UpdatePostInput!): UpdatePostPayload
}

type CreatePicturePayload {
  node: Picture!
  clientMutationId: String
}

input CreatePictureInput {
  content: RequiredPictureInput!
  clientMutationId: String
}

input RequiredPictureInput {
  url: String!
}

type UpdatePicturePayload {
  node: Picture!
  clientMutationId: String
}

input UpdatePictureInput {
  id: ID!
  content: PictureInput!
  options: UpdateOptionsInput
  clientMutationId: String
}

input PictureInput {
  url: String
}

input UpdateOptionsInput {
  replace: Boolean
  version: CeramicCommitID
}

\\"\\"\\"A Ceramic Commit ID\\"\\"\\"
scalar CeramicCommitID

type CreatePostPayload {
  node: Post!
  clientMutationId: String
}

input CreatePostInput {
  content: RequiredPostInput!
  clientMutationId: String
}

input RequiredPostInput {
  text: String!
}

type UpdatePostPayload {
  node: Post!
  clientMutationId: String
}

input UpdatePostInput {
  id: ID!
  content: PostInput!
  options: UpdateOptionsInput
  clientMutationId: String
}

input PostInput {
  text: String
}"
`;

exports[`graphql graphql:schema printing graphql schema succeeds with --readonly flag 1`] = `
"type Query {
  \\"\\"\\"Fetches an object given its ID\\"\\"\\"
  node(
    \\"\\"\\"The ID of an object\\"\\"\\"
    id: ID!
  ): Node
  account(id: DID!): CeramicAccount!
  viewer: CeramicAccount
}

\\"\\"\\"An object with an ID\\"\\"\\"
interface Node {
  \\"\\"\\"The id of the object.\\"\\"\\"
  id: ID!
}

type CeramicAccount {
  id: DID!
  isViewer: Boolean!
  data: CeramicAccountData!
}

\\"\\"\\"
A field whose value conforms to the standard DID format as specified in did-core: https://www.w3.org/TR/did-core/.
\\"\\"\\"
scalar DID @specifiedBy(url: \\"https://www.w3.org/TR/did-core/\\")

type CeramicAccountData {
  pictureCollection(
    \\"\\"\\"Returns the items in the list that come after the specified cursor.\\"\\"\\"
    after: String

    \\"\\"\\"Returns the first n items from the list.\\"\\"\\"
    first: Int

    \\"\\"\\"Returns the items in the list that come before the specified cursor.\\"\\"\\"
    before: String

    \\"\\"\\"Returns the last n items from the list.\\"\\"\\"
    last: Int
  ): PictureConnection
  postCollection(
    \\"\\"\\"Returns the items in the list that come after the specified cursor.\\"\\"\\"
    after: String

    \\"\\"\\"Returns the first n items from the list.\\"\\"\\"
    first: Int

    \\"\\"\\"Returns the items in the list that come before the specified cursor.\\"\\"\\"
    before: String

    \\"\\"\\"Returns the last n items from the list.\\"\\"\\"
    last: Int
  ): PostConnection
}

\\"\\"\\"A connection to a list of items.\\"\\"\\"
type PictureConnection {
  \\"\\"\\"Information to aid in pagination.\\"\\"\\"
  pageInfo: PageInfo!

  \\"\\"\\"A list of edges.\\"\\"\\"
  edges: [PictureEdge]
}

\\"\\"\\"Information about pagination in a connection.\\"\\"\\"
type PageInfo {
  \\"\\"\\"When paginating forwards, are there more items?\\"\\"\\"
  hasNextPage: Boolean!

  \\"\\"\\"When paginating backwards, are there more items?\\"\\"\\"
  hasPreviousPage: Boolean!

  \\"\\"\\"When paginating backwards, the cursor to continue.\\"\\"\\"
  startCursor: String

  \\"\\"\\"When paginating forwards, the cursor to continue.\\"\\"\\"
  endCursor: String
}

\\"\\"\\"An edge in a connection.\\"\\"\\"
type PictureEdge {
  \\"\\"\\"The item at the end of the edge\\"\\"\\"
  node: Picture

  \\"\\"\\"A cursor for use in pagination\\"\\"\\"
  cursor: String!
}

type Picture implements Node {
  id: ID!
  url: String!
}

\\"\\"\\"A connection to a list of items.\\"\\"\\"
type PostConnection {
  \\"\\"\\"Information to aid in pagination.\\"\\"\\"
  pageInfo: PageInfo!

  \\"\\"\\"A list of edges.\\"\\"\\"
  edges: [PostEdge]
}

\\"\\"\\"An edge in a connection.\\"\\"\\"
type PostEdge {
  \\"\\"\\"The item at the end of the edge\\"\\"\\"
  node: Post

  \\"\\"\\"A cursor for use in pagination\\"\\"\\"
  cursor: String!
}

type Post implements Node {
  id: ID!
  text: String!
}"
`;
